### ISSUE TITLE
GOMAXPROCS as number of cores increases latency in ping-pong benchmark

### ISSUE THREAD

**Issue by apolcyn** [grpc/grpc-go#1014]
The current scenario config sets `core_limit = 0` for go client/server QPS workers. With this config, QPS workers set `GOMAXPROCS` to 32 (number of available cores). But for ping pong test there's only one stream. 
Seeing more than 5% drop in latency by setting `GOMAXPROCS` to 1 on both client and server.
But note the numbers taken here vary from noise within maybe ~5%.

profile of go client during ping pong test (some hints at high contention):
~20% cpu in `runtime.futex`
~10% cpu in `_raw_spin_unlock_irqrestore` kernel function
~5% cpu in `finish_task_switch` kernel function
too much thread sync with high GOMAXPROCS?

After setting `GOMAXPROCS` to 1:
~13% cpu in `runtime.futex`
~5% cpu in `_raw_spin_unlock_irqrestore` kernel function
~3.4% cpu in `finish_task_switch` kernel function

ballpark numbers from a few ping pong runs on 32-core to 32-core client to server (subject to noise):
c++ -> c++ latency: 172/202 us for 50/90%
go -> go latency: 204/239 us for 50/90% (with current config)

`GOMAXPROCS = 1` on server:
c++ -> go latency: 178/206 us for 50/90%
`GOMAXPROCS = 1` on client:
go -> c++ latency: 190/220 us for 50/90%

on both:
go -> go latency: 177/207 us 50/90% - almost at c++ 

we can set this up in the benchmark by setting `core_limit = 1` in go QPS worker scenario config. this worth doing?

cc @iamqizhao @ctiller

**Comment by xiang90** [comment]
> we can set this up in the benchmark by setting core_limit = 1 in go QPS worker scenario config. this worth doing?

I do not think we should do this. This seems like a heck way to improve benchmark. For a lot of applications, we cannot really set core_limit to 1. I believe people are more interested in seeing the benchmark in a more default like setup.

**Comment by apolcyn** [comment]
@xiang90 makes sense, but just to be clear, setting `core_limit = 1` in the scenario config is actually just setting `GOMAXPROCS` to 1 (the default for go). We're currently breaking defaults by setting `GOMAXPROCS = 32` (on 32 core machine), which seems to be hurting performance for in this one test.
Though I'm not sure if a high `GOMAXPROCS` on a server might be common if not default.

**Comment by xiang90** [comment]
> (the default for go).

The default is max, not 1 since go1.5. See https://golang.org/doc/go1.5#introduction

**Comment by apolcyn** [comment]
I see, sorry I was mistaken here, closing as this is probably not going to be commonly done

**Comment by vjpai** [comment]
I suggest reopening and doing what you propose. These are microbenchmarks, not application-level benchmarks. The goal of ping-pong latency testing is to know the ultimate limit of what's achievable with a given API, not the practical reality. The latter is where other tests come into play.

**Comment by xiang90** [comment]
@vjpai I disagree. If you disable gc, ping-pong can run faster, should we also do that? we should not change runtime to make it run slightly faster. Also that option is going to go away, see godoc if you're not familiar with it.

**Comment by iamqizhao** [comment]
I think we should use Go default as the default and ppl can change it when they run the benchmark. It is not justified that 1 is the value to achieve the ultimate limit of latency. Even if it is right now, when the environment change (e.g., machine is upgraded or golang gets a new version) it might not be true any more. We do not have an automatic procedure to adjust its value.

**Comment by vjpai** [comment]
I'm going to suggest reopening this issue again. Not saying that it should be done, but I think we're missing the point of microbenchmarking. @xiang90 - disabling GC is not a realistic operating point for Go, but tuning the GC certainly is. Information that we discover from doing so should be used both to drive our own memory usage and to feed information up to the Go gc team. With regard to GOMAXPROCS, "This call will go away when the scheduler improves", and we don't know when that will be. Experimenting with this value gives us insights both into our own system and as potential feedback to the Go scheduling team. Likewise @iamqizhao - @apolcyn is noticing this effect today. Sure, let the Go default be the default, but report the number from the more effectively tuned version so that we're focusing on what gRPC can do, not the secondary effects created by the scheduler and other system vagaries that we actually have some control over. It certainly does not mean that it will continue in the future, but it is something that can be tested consistently. In the C++ performance benchmarking world, we have introduced daily sweep tests to experiment with various parameters; this could just as well fit into that.

**Comment by xiang90** [comment]
@vjpai 

> but tuning the GC certainly is

What do you mean be tuning GC? To get the max performance for the pingpong test? Or to use mircobench to understand the system?
Do you have any mechinism to do this for the ping pong test yet? 

> With regard to GOMAXPROCS

How do you know PROCS=1 works the best? Why not PROCS=2? Even for the ping-pong test, how do you know it can only utilize one core?

Further more, what the original issue suggested is to set the maxproc to 1, which seems to increase 5% for a specific go version on a specific machine. I still strongly disagree with that.

Also in your last reply, you changed your position from

> The goal of ping-pong latency testing is to know the ultimate limit of what's achievable with a given API, not the practical reality.

to 

> Experimenting with this value gives us insights both into our own system and as potential feedback to the Go scheduling team.

They are VERY DIFFERENT thing. I think you already confused yourself.

If you want to try different go runtime paras to figure the limit of gRPC-go or the limit of go runtime, that is probably find. As @iamqizhao mentioned, we need to the infrastructure to enable that. Or the best way is to leave this as default value instead of a vague value that can be changed eventually.

**Comment by xiang90** [comment]
In summary, I do not think we should tune runtime to achieve a higher bench result. If we want to understand the system better with microbench, we can tune paras. But we need to have the mechinism first instead of changing to a single value that seems to work the best.

**Comment by apolcyn** [comment]
Just since more runtime tuning is under discussion here, I'll add that experimentally I have seen good increases in throughput by raising GOGC in benchmarks - understand this is definitely under discussion but there's a request that does this in https://github.com/grpc/grpc/pull/8448 (tuning params provided by the benchmark runners in these requests)

**Comment by xiang90** [comment]
@apolcyn Sure. We should finish the discussion about tuning paras before setting this to a fixed value of 1. Also we need to constantly evaluate these values.

**Comment by ctiller** [comment]
Our standard for these benchmarks has always been that it's ok to tune any
knob that a user could *reasonably* be expected to tune in order to
demonstrate the best that we can do.

I'd argue GOGC and GOMAXPROCS certainly fit this.

As far as configuration goes, I'd place the standard as being someone
familiar with the benchmarks could discover the configuration changes made
per benchmark (ideally we'd call it out in the benchmarking logs).

I agree we need to periodically reverify these things. For C/C++ we'll be
doing sweeps of parameter spaces automatically on a reduced schedule from
our normal benchmarking efforts. As we notice better configurations we'll
certainly change to them. Alternatives would be to file bugs to revalidate
say... Once per quarter.

This parameter space searching is important work: I hope early in the new
year we can start writing performance optimization guides for all of our
stacks, and data points like 'switching to one core for ping pong bound
workloads' are valuable for those efforts.

**Comment by xiang90** [comment]
@ctiller 

> Our standard for these benchmarks has always been that it's ok to tune any
knob that a user could *reasonably* be expected to tune in order to
demonstrate the best that we can do.

What do you mean by

> a user could *reasonably* be expected to tune? 

Do you mean that users can also make the tune easily (in this case setting go proc == simply adding a evn var)? Or they think the tuning itself is reasonable (setting go proc = 1 is reasonable to users)?

**Comment by ctiller** [comment]
It's an intentionally loose definition: I usually take it as meaning would
a power user tuning an application that's expected to run in production
turn to this knob.

For me, that draws a line between disabling gc - which is probably
interesting for benchmark analysis but not for real systems, and tuning
core and thread counts and gc frequency and thresholds - which are more
likely to be early tools for power users to reach for to find that last xx%
for their application.

**Comment by apolcyn** [comment]
@xiang90 I'd add that there are published suggestions to tune these - I believe these knobs are meant to be used.

from https://software.intel.com/en-us/blogs/2014/05/10/debugging-performance-issues-in-go-programs
```
So it can make sense to set GOGC to a higher value (200, 300, 500, etc) if you can afford extra memory consumption.
```

from https://blog.golang.org/go15gc:
```
If you want to lower the total time spent in GC, increase GOGC
```

**Comment by xiang90** [comment]
>   an application that's expected to run in production

Changing to number of core to 1 is not realistic to the majority of Go production code I believe. I work on etcd, and it uses gRPC. We will never suggest any production users to change Core to 1 for gRPC. /cc @tamird I think Cockroachdb falls in the same bucket.

If you think it is realistic to change Go proc =1 for gRPC, give me examples.

Go is not like C++, where I can play around freely with the threading library and interact with raw OS.

**Comment by xiang90** [comment]
@apolcyn 

I understand we can tune go GC. That is a reasonable thing to do for a lot of cases.

**Comment by ctiller** [comment]
If you were working on an application that connected to a single peer and
exchanged messages would you set max procs to 1? Such applications exist,
and we need an idea of how grpc performs for them also.

**Comment by xiang90** [comment]
> If you were working on an application that connected to a single peer and exchanged messages would you set max procs to 1

Probably not. Making procs to N can make gc run concurrently with application. Also my application might want to process the data in async manner. 

Also do you have any real world example? Are these application the large majority?

It seems to me that you are simply making it up to justify your decision.

**Comment by ctiller** [comment]
I know of plenty of co-process applications (not all written in Go, but
they could conceivably be) that don't need a high thread count, but do
depend on low latency: proxys such as Lyft's Envoy spring to mind (and such
use cases are becoming more common so far as I can tell).

**Comment by xiang90** [comment]
>  don't need a high thread count

I think we are talking about exact 1 here and a go program.

But I am curious that Lyft's proxy only handle at most one request concurrently with 0 pipelining? They set java thread to exact 1?

/cc @mattklein123 is this true?

**Comment by carl-mastrangelo** [comment]
Adding fuel to the fire:  Java could reasonably tune GC knobs, but we mostly don't for benchmarks.  The machines that these benchmarks run on have copious amount of ram, which would let us avoid doing GC for very long periods of time.  This would artificially inflate Java's throughput, but would be outside the parameter hull of reasonable settings.

**Comment by mattklein123** [comment]
Is what true? Can you summarize the question?

**Comment by tamird** [comment]
Was there a resolution here?

**Comment by menghanl** [comment]
We are not setting `GOMAXPROCS` specifically in our benchmark, and are working on general performance improvements. So there's no point keeping this issue open now.
